rasterio/env.py

```diff
@@ -197,5 +198,8 @@ def delenv():
 def ensure_env(f):
     """A decorator that ensures an env exists before a function
     calls any GDAL C functions."""
-    defenv()
-    return f
+    @wraps(f)
+    def wrapper(*args, **kwds):
+        with Env(WITH_RASTERIO_ENV=True):
+            return f(*args, **kwds)
+    return wrapper
```

Comment:
Here's the crux of the PR: the `ensure_env` decorator ensures that a function is called in the context of a `Env`, mostly to get its GDAL error handling properly configured.